### PR TITLE
[Dist] enable to partition many chunks into less partitions via pipeline

### PR DESCRIPTION
**Pull request by Rhett-Ying**
## Description

`num_chunks` could be larger/equal than `num_parts`. Target for https://github.com/dmlc/dgl/issues/4362
## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: tools/distpartitioning/convert_partition.py [hunk 1]
```diff
@@ -95,11 +89,13 @@ def create_dgl_object(graph_name, num_parts, \
         map between edge type(string)  and edge_type_id(int)
     """
     #create auxiliary data structures from the schema object
-    ntid_dict, global_nid_ranges = get_idranges(schema[constants.STR_NODE_TYPE], 
-                                    schema[constants.STR_NUM_NODES_PER_CHUNK])
+    _, global_nid_ranges = get_idranges(schema[constants.STR_NODE_TYPE],
+                                    schema[constants.STR_NUM_NODES_PER_CHUNK],
+                                    expected_num_chunks=num_parts)
```
Comment (jermainewang) [inline tools/distpartitioning/convert_partition.py:94]: Have you formatted the code with autopep8?

File: tests/tools/test_dist_part.py [hunk 2]
```diff
@@ -152,11 +152,12 @@ def rand_edges(num_src, num_dst, num_edges):
                 assert feat_array.shape[0] == num_edges[etype] // num_chunks
 
         # Step1: graph partition
+        num_parts = 2
```
Comment (jermainewang) [inline tests/tools/test_dist_part.py:155]: Let's add more tests instead of having one test case. Can you add some corner cases too? e.g., num_parts=1, or num_chunks=1, or large num_chunks but small num_parts, or num_chunks is not divisible by num_parts, etc.

File: tools/distpartitioning/data_shuffle.py [hunk 3]
```diff
@@ -550,7 +552,7 @@ def gen_dist_partitions(rank, world_size, params):
     ntypes_gnid_range_map = get_gnid_range_map(node_tids)
     node_data, rcvd_node_features, rcvd_global_nids, edge_data  = \
                     exchange_graph_data(rank, world_size, node_features, node_feat_tids, \
-                                        edge_data, id_lookup, ntypes_ntypeid_map, ntypes_gnid_range_map, \
+                                        edge_data, id_lookup, ntypes_gnid_range_map, \
```
Comment (jermainewang) [inline tools/distpartitioning/data_shuffle.py:555]: `\` is not needed in multiline parenthesis.

File: tools/distpartitioning/data_shuffle.py [hunk 4]
```diff
@@ -288,7 +289,7 @@ def exchange_node_features(rank, world_size, node_feature_tids, ntype_gnid_map,
     return own_node_features, own_global_nids
 
 def exchange_graph_data(rank, world_size, node_features, node_feat_tids, edge_data,
-        id_lookup, ntypes_ntypeid_map, ntypes_gnid_range_map, ntid_ntype_map, schema_map):
+        id_lookup, ntypes_gnid_range_map, ntid_ntype_map, schema_map):
```
Comment (jermainewang) [inline tools/distpartitioning/data_shuffle.py:292]: Any new argument needs proper docstring.

File: tools/distpartitioning/dataset_utils.py [hunk 5]
```diff
@@ -224,26 +226,38 @@ def get_dataset(input_dir, graph_name, rank, world_size, schema_map):
...
+        for idx in read_list[rank]:
+            edge_file = edge_info[idx]
+            if not os.path.isabs(edge_file):
+                edge_file = os.path.join(input_dir, edge_file)
+            logging.info(f'Loading edges of etype[{etype_name}] from {edge_file}')
+            data_df = csv.read_csv(edge_file,
+                        read_options=pyarrow.csv.ReadOptions(autogenerate_column_names=True), 
+                        parse_options=pyarrow.csv.ParseOptions(delimiter=' '))
+            df_0.append(data_df['f0'].to_numpy())
+            df_1.append(data_df['f1'].to_numpy())
```
Comment (jermainewang) [inline tools/distpartitioning/dataset_utils.py:250]: Can we have better names than `'f0'` and `'f1'`?

Comment (jermainewang) [reply tools/distpartitioning/dataset_utils.py:250]: Same for `df_0` and `df_1`.

File: tools/distpartitioning/utils.py [hunk 6]
```diff
@@ -387,6 +381,8 @@ def get_idranges(names, counts):
         list of node/edge types as strings
     counts : list of lists
         each list contains no. of nodes/edges in a given chunk
+    expected_num_chunks : expected number of returned chunks
+        chunks are collapsed according to expected number of chunks
```
Comment (jermainewang) [inline tools/distpartitioning/utils.py:385]: - What goes after the colon symbol `:` should be the type of the argument.
- The description does not make sense to me. Please reword.

File: tools/distpartitioning/utils.py [hunk 7]
```diff
@@ -402,21 +398,31 @@ def get_idranges(names, counts):
...
         gnid_end += tid_ranges[-1][1]
 
         tid_dict[typename] = tid_ranges
         gid_dict[typename] = np.array([gnid_start, gnid_end]).reshape([1,2])
 
         gnid_start = gnid_end
+        orig_num_chunks = len(tid_start)
+
+    assert expected_num_chunks <= orig_num_chunks, \
+        'expected num_chunks should be less/euqual than original one.'
```
Comment (jermainewang) [inline tools/distpartitioning/utils.py:417]: What is "original one" here? The error message will not make any sense to the users.

**Review by jermainewang — CHANGES_REQUESTED** [review]

File: tools/dispatch_data.py [hunk 8]
```diff
@@ -75,6 +78,8 @@ def main():
     parser.add_argument('--ssh-port', type=int, default=22, help='SSH Port.') 
     parser.add_argument('--process-group-timeout', type=int, default=1800,
                         help='timeout[seconds] for operations executed against the process group')
+    parser.add_argument('--num-parts', type=int, default=0,
+                        help='number of target partitions. If not specified, number of chunks are used.')
```
[the remark this answers is not in the document]
Comment (Rhett-Ying) [reply tools/dispatch_data.py:84]: yes. a json file with single line is enough.

File: tools/dispatch_data.py [hunk 9]
```diff
@@ -40,7 +40,10 @@ def submit_jobs(args) -> str:
     with open(os.path.join(args.in_dir, schema_path)) as schema:
         schema_map = json.load(schema)
 
-    num_parts = len(schema_map["num_nodes_per_chunk"][0])
+    num_chunks = len(schema_map["num_nodes_per_chunk"][0])
+    if args.num_parts > num_chunks:
+        raise Exception('Number of partitions should be less/equal than number of chunks.')
+    num_parts = num_chunks if args.num_parts <= 0 else args.num_parts
```
[the remark this answers is not in the document]
Comment (Rhett-Ying) [reply tools/dispatch_data.py:46]: I don't think so

File: tools/distpartitioning/data_shuffle.py (hunk 4 above)
Comment (Rhett-Ying) [reply tools/distpartitioning/data_shuffle.py:294]: here, I just removed unused argument.

File: tools/dispatch_data.py [hunk 10]
```diff
@@ -33,16 +33,29 @@ def get_launch_cmd(args) -> str:
...
+        with open(partition_path) as part_meta:
+            num_parts = json.load(part_meta)['num_parts']
+    if num_parts > num_chunks:
+        raise Exception('Number of partitions should be less/equal than number of chunks.')
+
+    # verify ip_config
+    with open(args.ip_config, 'r') as f:
+        num_ips = len(f.readlines())
+        assert num_ips == num_parts, \
+            f'Number of lines in {args.ip_config} should equal num_parts[{num_parts}].'
```
Comment (jermainewang) [inline tools/dispatch_data.py:57]: The number of lines ... should be equal to ...

File: tools/partition_algo/random_partition.py [hunk 11]
```diff
@@ -17,6 +17,15 @@ def _random_partition(metadata, num_parts):
         parts = np.random.randint(0, num_parts, (n,))
         array_readwriter.get_array_parser(name='csv').write(ntype + '.txt', parts)
 
+def _dump_metadata(num_parts):
+    """
+    Dump metadata into file. For now, below metadata are dumped:
+    ``version``, ``num_parts``.
+    """
+    metadata = {'version': '0.0.1', 'num_parts': num_parts}
```
[the remark this answers is not in the document]
Comment (Rhett-Ying) [reply tools/partition_algo/random_partition.py]: added. pls take a look

**Comment by kylasa** [comment]
On a general level, This implementation ignores this case when merging different chunks to form a total number of chunks equal to no. of partitions required. Since chunks are treated first come first serve, consider the case where we have num_nodes_per_chunk = [10 ,10, 10, 10, 1] and no. of required partitions is set to `3`. If this is the case, this implementation will result in updated num_nodes_per_chunk = [20, 20, 1]. The same goes for edges as well. Now the load on each process is highly lop-sided. We can assume several such cases where the later processes will have smaller amount of load compared to earlier processes. 

For these above cases, if the nodes/edges per chunk is set to millions or billions then this will result in overly lop-sided processes when doing disk I/O. 

For now, we can treat this as a corner case, but at some point we should address this use-case. I think this is very important  unit test case.

**Comment by Rhett-Ying** [comment]
@kylasa Yes. You're right. Current implementation will hit load imbalance issue. I think one quick/simple fix for this issue is to split the array to `[ [10,10], [10, 1], [10] ]`. This will mitigate the imbalance though not the optimized one: in average. Average workload may introduces much logic, I think.

**Review by kylasa — APPROVED** [review]
Overall I agree with the changes... Please take appropriate actions...

File: tools/distpartitioning/utils.py [hunk 12]
```diff
@@ -85,15 +85,9 @@ def get_ntype_featnames(ntype_name, schema_map):
...
-        featnames = []
-        ntype_info = ntype_featdict[ntype_name]
-        for k, v in ntype_info.items(): 
-            featnames.append(k)
-        return featnames
-    else: 
-        return []
+    node_data = schema_map[constants.STR_NODE_DATA]
+    feats = node_data.get(ntype_name, {})
+    return [feat for feat in feats]
```
[the remark this answers is not in the document]
Comment (Rhett-Ying) [reply tools/distpartitioning/utils.py:90]: I think it behaves the same as previous one. if `ntype_name` not in `node_data`, empty array will be returned.

File: tools/partition_algo/base.py [hunk 13]
```diff
@@ -0,0 +1,52 @@
+from typing import Optional
+import pydantic as dt
+import json
+from dgl import DGLError
+
+class PartitionMeta(dt.BaseModel):
+    """ Describe fields of partition metadata.
```
Comment (jermainewang) [inline tools/partition_algo/base.py:7]:
```
"""Metadata that describes the partition assignment results.

Regardless of the choice of partitioning algorithm, a metadata JSON file will be created
in the output directory which includes the meta information of the partition algorithm.

To generate a metadata JSON:

<give an example>

To read a metadata JSON:

<give an example>
"""
```

File: tools/partition_algo/base.py [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+    """
+    version: Optional[str] = '1.0.0'
+    num_parts: int
```
Comment (jermainewang) [inline tools/partition_algo/base.py:10]: Add a field "algo_name"

Comment (jermainewang) [reply tools/partition_algo/base.py:10]: Also add comment for each field.

File: tools/partition_algo/base.py [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+    -------
+    PartitionMeta
+        The partition metadata.
+    """
+    with open(meta_file) as f:
+        try:
+            part_meta = PartitionMeta(**(json.load(f)))
+        except dt.ValidationError as e:
+            print(
+                "Details of pydantic.ValidationError:\n{}".format(e.json()))
```
Comment (jermainewang) [inline tools/partition_algo/base.py:43]: "Invalid partition metadata JSON. Error details:\{}"

File: tools/partition_algo/base.py [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+            raise DGLError(
+                "Validation Error for YAML fields. Details are shown above.")
```
Comment (jermainewang) [inline tools/partition_algo/base.py:45]: - Why not putting the error message here?
- JSON not YAML

File: tools/dispatch_data.py [hunk 17]
```diff
@@ -33,16 +34,29 @@ def get_launch_cmd(args) -> str:
...
     with open(os.path.join(args.in_dir, schema_path)) as schema:
         schema_map = json.load(schema)
 
-    num_parts = len(schema_map["num_nodes_per_chunk"][0])
     graph_name = schema_map["graph_name"]
 
+    # retrieve num_parts
+    num_chunks = len(schema_map["num_nodes_per_chunk"][0])
+    num_parts = num_chunks
+    partition_path = os.path.join(args.partitions_dir, "partition.json")
```
Comment (jermainewang) [inline tools/dispatch_data.py:47]: Rename it to `metadata.json`?

**Review by jermainewang — CHANGES_REQUESTED** [review]

File: tools/dispatch_data.py (hunk 17 above)
Comment (Rhett-Ying) [reply tools/dispatch_data.py:47]: I think `metadata.json` is used abuse, though we could distinguish according to the base dir.

Comment (Rhett-Ying) [reply tools/dispatch_data.py:47]: `partition_meta.json` ?

**Review by jermainewang — APPROVED** [review]
